### ISSUE TITLE
Deprecation of ldapjs

### ISSUE THREAD

**Issue by wesley3295** [shaozi/ldap-authentication#63]
Any ideas on what we can do in replacement of ldapjs? This package is something my company relies on heavily and with it not being maintained anymore throws a huge plot twist into our current work with our client. Is there another open source ldap package that can replace ldapjs or something I/we can contribute to?

**Comment by shaozi** [comment]
It is a very sad that some stupid coward threatened the maintainer and now the project is archived. 

So far I have not found anything yet. We should find some people to take the project over since it is open source.

**Comment by mattkrins** [comment]
I don't know of any alternatives sadly... It was one of, if not the **only** decent ldap packages for node.
I've never seen such an unhinged deprecation readme before.
It was clearly a troll/spam email but I guess the author wanted an excuse to stop maintaining it.

**Comment by stalkerg** [comment]
I think we can easily port it to https://github.com/jjhoughton/napi-ldap or make it abstract.

**Comment by stalkerg** [comment]
or maybe it will be better just to reimplement such features by Rust?  https://github.com/inejge/ldap3

**Comment by shaozi** [comment]
> or maybe it will be better just to reimplement such features by Rust?  https://github.com/inejge/ldap3 

I like this idea. Any pointers on how?

**Comment by stalkerg** [comment]
> I like this idea. Any pointers on how?

I think the easier way to make binding by this
https://github.com/neon-bindings/neon

It should be relatively easy because we do not need many features. 
If you want, I can try/help next week. 

PS Of course, true JS implementation has some benefits, but we do not have easy solutions in the current situation.

**Comment by stalkerg** [comment]
It seems like we are using one `filter`: https://github.com/ldapjs/filter. Do we need it? As I understand, it's IR, and in our case, `(${userAttr}=${userName})` it's overkill.

**Comment by stalkerg** [comment]
For the rust idea, I think it will be easier to implement high-level functions  - `_searchUser` and `_searchUserGroups`.

**Comment by shaozi** [comment]
> It seems like we are using one `filter`: https://github.com/ldapjs/filter. Do we need it? As I understand, it's IR, and in our case, `(${userAttr}=${userName})` it's overkill.

What is `IR`? Now I look into the code, it does seem over complicated in the `ldapjs/filter` implementation. But I wonder if we simply use string substitute, do we need to do any pre-processing of the user input to ease any security injection concern? Of course, I am not sure the `filter` lib does that, but I just assume it does something.

**Comment by shaozi** [comment]
The neon-bindings is a really nice lib. I searched and there is another one called [napi-rs](https://github.com/napi-rs/napi-rs). Have you compared these two?

**Comment by stalkerg** [comment]
> What is IR?

https://en.wikipedia.org/wiki/Intermediate_representation 

>  do we need to do any pre-processing of the user input to ease any security injection concern?

Yes, sure, but it is trivial; all literals should be escaped. Rust lib already supports it: https://docs.rs/ldap3/0.11.3/ldap3/fn.ldap_escape.html

> The neon-bindings is a really nice lib. I searched and there is another one called [napi-rs](https://github.com/napi-rs/napi-rs). Have you compared these two?

Yeah, it seems like napi-rs even better! 
https://napi.rs/docs/more/neon

**Comment by shaozi** [comment]
Cool let me try it out next week and see how is it going.

**Comment by stalkerg** [comment]
do we have any updates on a such hot topic?

**Comment by kedom1337** [comment]
Has anybody already evaluated [ldapts](https://github.com/ldapts/ldapts) as an alternative?

**Comment by shaozi** [comment]
> Has anybody already evaluated [ldapts](https://github.com/ldapts/ldapts) as an alternative?

This one looks nice. As I said, I will keep v3 as pure js/ts, so this one may fit there.

**Comment by stalkerg** [comment]
Yeah, ldapts also looks good, maybe even easier for a short term.

**Comment by pdvcs** [comment]
I wanted to express my appreciation to @shaozi for looking after this library and for looking for a path forward. Thank you!

**Comment by ryanm101** [comment]
Possible replacement: https://github.com/ldapts/ldapts?tab=readme-ov-file#authenticate-example

**Comment by erickweil** [comment]
I needed to autenticate with ldap and after frustating search of libraries (lots are decomissioned/abandoned), decided to use ldapts directly, and used your project as high level guideline on how ldap auth works. I managed to reverse engineer and adapt for my use case but using ldapts instead as underlying library.

In order to help this project, considering this open issue, I made a pull request with the changes needed, which seem to be not that much: [#78](https://github.com/shaozi/ldap-authentication/pull/78)

**Comment by shaozi** [comment]
@erickweil I added some review comments in #78 
Would you take a look and fix them?

**Comment by shaozi** [comment]
I will close this issue. Thank @erickweil for providing the migration to ldapts. fixed by #78
